### 1. Import Libraries

In [4]:
import pandas as pd 
import numpy as np
import os 
from pathlib import Path

### 2. Reading the Data

In [10]:
import os
import pandas as pd

file_names = ["P_banphim.csv", "banvedientu.csv", "P_chuot.csv", "danamthanh.csv", "dienthoai_phukien.csv", "P_gameconsole.csv",
              "laptop.csv", "P_linhkienmaytinh.csv", "pc.csv", "maynghenhac.csv", "P_micro.csv", "tainghe.csv",
              "tivi.csv", "sac.csv", "maytinhbang.csv","machdientu.csv","tainghecoday.csv"]

dataframes = []
categories = ["Bàn phím", "Bản vẽ điện tử", "Chuột", "Âm thanh", "Điện thoại phụ kiện", "Game console",
              "Laptop", "Linh kiện máy tính", "PC", "Máy nghe nhạc", "Micro", "Tai nghe",
              "Tivi", "Sạc", "Máy tính bảng", "Mạch điện tử", "Tai nghe"]

csv_folder = "E:\MY CAREER\Internship Project\EDA\dataset\shopee\Product"

for file_name, category in zip(file_names, categories):
    path = os.path.join(csv_folder, file_name)
    try:
        df = pd.read_csv(path)
        df["category"] = category  # Thêm cột "category" với giá trị tương ứng
        dataframes.append(df)
    except pd.errors.ParserError as e:
        print(f"Error reading file {file_name}: {e}")

combined_df = pd.concat(dataframes, ignore_index=True)
combined_df.head()


Unnamed: 0       itemid     shopid  \
0           0   1546899006   93922606   
1           1   6989181845  268182642   
2           2   3066084859   45654584   
3           3  20259729410   12208293   
4           4  17817240547  697719850   

                                                name     brand  subcatid  \
0  Bàn phím giả cơ gaming kèm chuột có dây chế độ...  No Brand    100644   
1  Bàn Phím Máy Tính Gaming RGB SIDOTECH LDK V4 P...       NaN    100644   
2  Bàn phím máy tính gaming G21, bàn phím giả cơ ...       NaN    100644   
3  Bàn Phím Cơ Bluetooth Không Dây K68BT Trục Cơ ...       NaN    100644   
4  bàn phím cơ rgb led K550 LEAVEN Giao hàng 24 g...       NaN    100644   

   cb_option  cmt_count    rating       ctime  ...        price  discount  \
0          0      28766  4.753601  1538622802  ...   4100000000       34%   
1          0       4438  4.775372  1621337118  ...  12400000000       50%   
2          0       2475  4.766680  1624443365  ...   3500000000       37%   
3          0       1060  4.893396  1667447533  ...  48900000000       39%   
4          0       4446  4.826811  1646292261  ...  31300000000       50%   

    stock  show_free_shipping    shop_location  shopee_verified  \
0  244373                True           Hà Nội             True   
1     250                True           Hà Nội            False   
2     448                True  TP. Hồ Chí Minh             True   
3      40               False           Hà Nội             True   
4  225143               False           Hà Nội             True   

  can_use_bundle_deal  can_use_cod   catid  category  
0               False         True  110360  Bàn phím  
1               False         True  110360  Bàn phím  
2               False         True  110360  Bàn phím  
3               False         True  110360  Bàn phím  
4               False         True  110360  Bàn phím  

[5 rows x 25 columns]

In [12]:
combined_df.shape

(16183, 25)

In [13]:
combined_df.columns

Index(['Unnamed: 0', 'itemid', 'shopid', 'name', 'brand', 'subcatid',
       'cb_option', 'cmt_count', 'rating', 'ctime', 'flag', 'historical_sold',
       'is_adult', 'is_official_shop', 'liked_count', 'price', 'discount',
       'stock', 'show_free_shipping', 'shop_location', 'shopee_verified',
       'can_use_bundle_deal', 'can_use_cod', 'catid', 'category'],
      dtype='object')

In [14]:
columns= ['itemid', 'shopid', 'name', 'brand', 'subcatid','category',
       'cb_option', 'cmt_count', 'rating', 'ctime', 'flag', 'historical_sold',
       'is_adult', 'is_official_shop', 'liked_count', 'price', 'discount',
       'stock', 'show_free_shipping', 'shop_location', 'shopee_verified',
       'can_use_bundle_deal', 'can_use_cod']
new_df = combined_df[columns]
new_df.head()

itemid     shopid                                               name  \
0   1546899006   93922606  Bàn phím giả cơ gaming kèm chuột có dây chế độ...   
1   6989181845  268182642  Bàn Phím Máy Tính Gaming RGB SIDOTECH LDK V4 P...   
2   3066084859   45654584  Bàn phím máy tính gaming G21, bàn phím giả cơ ...   
3  20259729410   12208293  Bàn Phím Cơ Bluetooth Không Dây K68BT Trục Cơ ...   
4  17817240547  697719850  bàn phím cơ rgb led K550 LEAVEN Giao hàng 24 g...   

      brand  subcatid  category  cb_option  cmt_count    rating       ctime  \
0  No Brand    100644  Bàn phím          0      28766  4.753601  1538622802   
1       NaN    100644  Bàn phím          0       4438  4.775372  1621337118   
2       NaN    100644  Bàn phím          0       2475  4.766680  1624443365   
3       NaN    100644  Bàn phím          0       1060  4.893396  1667447533   
4       NaN    100644  Bàn phím          0       4446  4.826811  1646292261   

   ...  is_official_shop  liked_count        price  discount   stock  \
0  ...             False        14312   4100000000       34%  244373   
1  ...              True         3325  12400000000       50%     250   
2  ...             False         1134   3500000000       37%     448   
3  ...             False         1632  48900000000       39%      40   
4  ...             False         3141  31300000000       50%  225143   

   show_free_shipping    shop_location  shopee_verified  can_use_bundle_deal  \
0                True           Hà Nội             True                False   
1                True           Hà Nội            False                False   
2                True  TP. Hồ Chí Minh             True                False   
3               False           Hà Nội             True                False   
4               False           Hà Nội             True                False   

  can_use_cod  
0        True  
1        True  
2        True  
3        True  
4        True  

[5 rows x 23 columns]

### 3. Data Cleaning & Exploring

In [15]:
num_duplicates = new_df.duplicated(subset='itemid').sum()
num_duplicates

348

Co 348 san pham bi trung lap the nen toi se tien hanh xoa chung di

In [16]:
new_df.drop_duplicates(subset='itemid', inplace=True, keep = 'last')
num_duplicates = new_df.duplicated(subset='itemid').sum()
num_duplicates

C:\Users\Rauciss\AppData\Local\Temp\ipykernel_19176\2519479065.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df.drop_duplicates(subset='itemid', inplace=True, keep = 'last')


0

In [17]:
new_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 15835 entries, 2 to 16182
Data columns (total 23 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   itemid               15835 non-null  int64  
 1   shopid               15835 non-null  int64  
 2   name                 15835 non-null  object 
 3   brand                3889 non-null   object 
 4   subcatid             15835 non-null  int64  
 5   category             15835 non-null  object 
 6   cb_option            15835 non-null  int64  
 7   cmt_count            15835 non-null  int64  
 8   rating               15835 non-null  float64
 9   ctime                15835 non-null  int64  
 10  flag                 15835 non-null  int64  
 11  historical_sold      15835 non-null  int64  
 12  is_adult             15835 non-null  bool   
 13  is_official_shop     15835 non-null  bool   
 14  liked_count          15835 non-null  int64  
 15  price                15835 non-null 

Nhan thay brand, discount, shop_location co cac gia tri null nen chung ta kiem tra va tien hanh xu ly gia tri thieu


In [18]:
brand = new_df['brand'].unique()
b = brand.tolist()
print(b)
len(brand)

[nan, 'No Brand', 'MOFII', 'logitech', 'DAREU', 'Newmen', 'No brand', 'Fuhlen', 'Electronic', 'E-DRA', '0', 'HP', 'R8', 'NoBrand', 'DELL', 'OEM', 'Prolink', 'XSmart', 'Free Wolf', 'LIMEIDE', 'MOFii', 'Meetion', 'E-Blue', 'None', 'BOSSTON', 'AULA', 'Onikuma', 'ASSASSINS', 'Logitech', 'PHILIPS', 'Ajazz', 'Royal Kludge', 'oem', 'MIXIE', 'LANGTU', 'ONEKA', 'iMice', 'Lofree', 'Acer', 'hoco.', 'Toshiba', 'Xiaomi', 'Wacom', 'HUION', 'XP-Pen', 'Bluelans', 'Sido Tech', 'Forter', 'Limeipe', 'Genius', 'Lenovo', 'Corsair', 'RAZER', 'HXSJ', 'Inphic', 'A4TECH', 'kebidumei', 'ROBOT', 'Mitsumi', 'ARS', 'JBL', 'Sony', 'SONY', 'Boston', 'BMB', 'Jarguar Suhyoung', 'Avector', 'ONTEK', 'Bose', 'ONTEKCO', 'BOSE', 'Arirang', 'LG', 'MICROLAB', 'audio-technica', 'Samsung', 'DIDO', 'Ugreen', 'microlab', 'ACNOS', 'SoundMax', 'Google', 'MIFA', 'Marshall', 'TRONSMART', 'harman/kardon', 'Fenda', 'RHM', 'KIOMIC', 'AUDIOENGINE', 'Bang & Olufsen', 'Google Home', 'Klipsch', 'Yayusi', 'JBZ', 'KOVIT', 'PKCB', 'CROWN', 'S

297

Có tới 300 brands song có nhiều giá trị trùng lặp: NoBrand, nobrand, sony, SONY,....

In [19]:
new_df["brand"] = new_df["brand"].str.lower().str.strip()
brand = new_df['brand'].unique()
len(brand)

C:\Users\Rauciss\AppData\Local\Temp\ipykernel_19176\3100919109.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df["brand"] = new_df["brand"].str.lower().str.strip()


264

In [20]:
sorted_brands = new_df["brand"].sort_values().unique()
e = sorted_brands.tolist()
print(e)

['0', '2good', 'a4tech', 'acer', 'acnos', 'ajazz', 'akai', 'akg', 'alcatel one touch idol', 'altec lansing', 'amazon', 'amazon fire 8', 'amoi', 'anker', 'apple', 'arirang', 'ariston', 'ars', 'asano', 'asanzo', 'asrock', 'assassins', 'asus', 'asus zenpad c', 'audio-technica', 'audioengine', 'aula', 'avector', 'avermedia', 'bailli', 'bang & olufsen', 'baseus', 'basike', 'belkin', 'benjie', 'benq', 'bitmain', 'black shark', 'blackberry', 'bluelans', 'bmax', 'bmb', 'borofone', 'bosch', 'bose', 'bosston', 'boston', 'byz', 'c.o.k', 'calcutta', 'camelion', 'casper', 'chint', 'chuwi', 'chuwi hi10pro.', 'clip tv box', 'comet', 'coocaa', 'coolman sunshine', 'corsair', 'craven', 'crossman', 'crown', 'dareu', 'darling', 'dell', 'devialet', 'dido', 'divoom', 'dobo', 'doublepow', 'e-blue', 'e-dra', 'earldom', 'edifier', 'elecom', 'electronic', 'energizer', 'essager', 'fenda', 'fengzhilife', 'fire hd 10 2017', 'foomee', 'forter', 'free wolf', 'fuhlen', 'fujitsu', 'galaxy store', 'genius', 'gigabyte',

In [21]:
new_df["brand"] = new_df["brand"].str.replace(".", "")
new_df["brand"] = new_df["brand"].replace({
    'intel™': 'intel',
    'nobrand': 'no brand',
    'none': 'no brand',
    'nan': 'no brand',
    'rock space': 'rockspace'
})
len(new_df["brand"].unique())


C:\Users\Rauciss\AppData\Local\Temp\ipykernel_19176\40797381.py:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  new_df["brand"] = new_df["brand"].str.replace(".", "")
C:\Users\Rauciss\AppData\Local\Temp\ipykernel_19176\40797381.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df["brand"] = new_df["brand"].str.replace(".", "")
C:\Users\Rauciss\AppData\Local\Temp\ipykernel_19176\40797381.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://

258

In [22]:
new_df['brand'].fillna('no brand')

2        no brand
5        no brand
6        no brand
7        no brand
9        no brand
           ...   
16178    no brand
16179    no brand
16180    no brand
16181    no brand
16182    no brand
Name: brand, Length: 15835, dtype: object

In [23]:
new_df["brand"] = new_df["brand"].fillna("no brand")

C:\Users\Rauciss\AppData\Local\Temp\ipykernel_19176\572235798.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df["brand"] = new_df["brand"].fillna("no brand")


In [24]:
new_df["discount"] = new_df["discount"].str.replace("%", "")
new_df["discount"] = pd.to_numeric(new_df["discount"], errors="coerce").fillna(0)


C:\Users\Rauciss\AppData\Local\Temp\ipykernel_19176\1357669426.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df["discount"] = new_df["discount"].str.replace("%", "")
C:\Users\Rauciss\AppData\Local\Temp\ipykernel_19176\1357669426.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df["discount"] = pd.to_numeric(new_df["discount"], errors="coerce").fillna(0)


In [25]:
new_df['shop_location'].value_counts()

Hà Nội             6079
TP. Hồ Chí Minh    4728
Nước ngoài         2465
Hải Phòng           271
Hải Dương           212
                   ... 
Kon Tum               2
Bắc Kạn               2
Hàn Quốc              2
Cao Bằng              2
Hong Kong             1
Name: shop_location, Length: 66, dtype: int64

In [26]:
new_df['shop_location'] = new_df['shop_location'].fillna('Hà Nội')
new_df["shop_location"] = new_df["shop_location"].replace({
    'Hàn Quốc': 'Nước ngoài',
    'Hong Kong': 'Nước ngoài'
})
len(new_df['shop_location'].unique())

C:\Users\Rauciss\AppData\Local\Temp\ipykernel_19176\1900039218.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df['shop_location'] = new_df['shop_location'].fillna('Hà Nội')
C:\Users\Rauciss\AppData\Local\Temp\ipykernel_19176\1900039218.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df["shop_location"] = new_df["shop_location"].replace({


64

In [27]:
new_df['price'] = new_df['price'] / 100000 

C:\Users\Rauciss\AppData\Local\Temp\ipykernel_19176\2046214833.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df['price'] = new_df['price'] / 100000


In [28]:

new_df['ctime'] = pd.to_datetime(new_df['ctime'], unit='s')
new_df.head()

C:\Users\Rauciss\AppData\Local\Temp\ipykernel_19176\2598832837.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df['ctime'] = pd.to_datetime(new_df['ctime'], unit='s')


itemid      shopid                                               name  \
2   3066084859    45654584  Bàn phím máy tính gaming G21, bàn phím giả cơ ...   
5  20011286067   792250698  Set bàn phím không dây bluetooth và chuột hoạt...   
6   7134465302    63347190             Bàn phím căn phòng có dây JDKEE KB-816   
7  20991373250  1032884533  Chuột Và Bàn Phím Máy Tính MAKTO, Chuột M1 + B...   
9   1833183314   118899185  Bàn phím Mini K1000 USB - Full Box- Chất Lượng...   

      brand  subcatid  category  cb_option  cmt_count    rating  \
2  no brand    100644  Bàn phím          0       2475  4.766680   
5  no brand    100644  Bàn phím          0       2468  4.645867   
6  no brand    100644  Bàn phím          0        585  4.464957   
7  no brand    100644  Bàn phím          0          0  0.000000   
9  no brand    100644  Bàn phím          0       2873  4.767061   

                ctime  ...  is_official_shop  liked_count    price  discount  \
2 2021-06-23 10:16:05  ...             False         1134  35000.0      37.0   
5 2022-06-13 08:57:36  ...             False         4808  49000.0      50.0   
6 2020-05-18 07:33:52  ...             False          179  59000.0      13.0   
7 2023-07-31 02:18:47  ...             False            0  83000.0      50.0   
9 2019-01-13 06:58:39  ...             False          925  70000.0      46.0   

    stock  show_free_shipping    shop_location  shopee_verified  \
2     448                True  TP. Hồ Chí Minh             True   
5  207287               False  TP. Hồ Chí Minh             True   
6      14               False           Hà Nội             True   
7    3000               False         Nam Định            False   
9       7                True  TP. Hồ Chí Minh             True   

   can_use_bundle_deal can_use_cod  
2                False        True  
5                False        True  
6                False        True  
7                False        True  
9                False        True  

[5 rows x 23 columns]

In [30]:
new_df.to_csv('cleandata\shopee_products.csv')